File: capture/test/browser-v2/scope.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";

import {
  setUser,
  setTag,
  setRequestContext,
  getUser,
  getTags,
  getRequestContext,
  getBreadcrumbs,
  clearScope,
  redactBody,
  scrubUrl,
  scrubSecrets,
} from "../dist/scope.js";
import { addBreadcrumb } from "../dist/scope.js";

test("setUser does not include email", () => {
  clearScope();
  setUser("u1", "admin");
  const u = getUser();
  assert.equal(u.id, "u1");
  assert.equal(u.role, "admin");
  assert.equal(u.email, undefined);
});

test("setTag accumulates", () => {
  clearScope();
  setTag("region", "us-east-1");
  setTag("k8s_namespace", "prod");
  const tags = getTags();
  assert.equal(tags.region, "us-east-1");
  assert.equal(tags.k8s_namespace, "prod");
});

test("setRequestContext redacts headers", () => {
  clearScope();
  setRequestContext({
    method: "POST",
    url: "/x",
    headers: {
      Authorization: "Bearer abc",
      "X-Auth-Token": "leak",
      Accept: "application/json",
    },
  });
  const req = getRequestContext();
  assert.equal(req.headers.Authorization, "[REDACTED]");
  assert.equal(req.headers["X-Auth-Token"], "[REDACTED]");
  assert.equal(req.headers.Accept, "application/json");
});

test("redactBody scrubs nested secrets", () => {
  const safe = redactBody({
    user: "alice",
    password: "hunter2",
    nested: { api_key: "sk_xxx", deep: { token: "leak" } },
    items: [{ secret: "X", ok: "Y" }],
  });
  assert.equal(safe.password, "[REDACTED]");
  assert.equal(safe.nested.api_key, "[REDACTED]");
  assert.equal(safe.nested.deep.token, "[REDACTED]");
  assert.equal(safe.items[0].secret, "[REDACTED]");
  assert.equal(safe.items[0].ok, "Y");
});

test("breadcrumb ring caps at 30", () => {
  clearScope();
  for (let i = 0; i < 50; i++) addBreadcrumb({ category: "test", message: `crumb-${i}` });
  const crumbs = getBreadcrumbs();
  assert.equal(crumbs.length, 30);
  assert.equal(crumbs[0].message, "crumb-20");
  assert.equal(crumbs[29].message, "crumb-49");
});

test("scrubUrl redacts sensitive query params", () => {
  const out = scrubUrl("https://api.example.com/x?token=abc&page=2");
  assert.ok(out.includes("token=[REDACTED]"));
  assert.ok(out.includes("page=2"));
});

test("scrubSecrets removes bearer tokens", () => {
  const out = scrubSecrets("Authorization: Bearer abc123");
  assert.ok(out.includes("[REDACTED]"));
});

File: capture/java/src/test/java/com/inariwatch/capture/ScopeTest.java
package com.inariwatch.capture;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class ScopeTest {

    @BeforeEach
    void reset() { Scope.clear(); Breadcrumbs.clear(); }

    @Test
    void setUserDoesNotIncludeEmail() {
        Scope.setUser("u1", "admin");
        Map<String, Object> u = Scope.getUser();
        assertEquals("u1", u.get("id"));
        assertEquals("admin", u.get("role"));
        assertNull(u.get("email"));
    }

    @Test
    void redactRequestScrubsHeaders() {
        Map<String, Object> headers = new HashMap<>();
        headers.put("Authorization", "Bearer abc");
        headers.put("X-Auth-Token", "leak");
        headers.put("Accept", "application/json");
        Map<String, Object> req = new HashMap<>();
        req.put("method", "POST");
        req.put("url", "/x");
        req.put("headers", headers);
        Scope.setRequestContext(req);
        @SuppressWarnings("unchecked")
        Map<String, Object> out = (Map<String, Object>) Scope.getRequestContext().get("headers");
        assertEquals("[REDACTED]", out.get("Authorization"));
        assertEquals("[REDACTED]", out.get("X-Auth-Token"));
        assertEquals("application/json", out.get("Accept"));
    }

    @Test
    void breadcrumbRingCapsAt30() {
        for (int i = 0; i < 50; i++) Breadcrumbs.add("test", "crumb-" + i);
        assertEquals(30, Breadcrumbs.get().size());
        assertEquals("crumb-20", Breadcrumbs.get().get(0).get("message"));
        assertEquals("crumb-49", Breadcrumbs.get().get(29).get("message"));
    }

    @Test
    void breadcrumbsScrubBearerTokens() {
        Breadcrumbs.add("http", "GET /x with Authorization: Bearer abcdef");
        String msg = (String) Breadcrumbs.get().get(0).get("message");
        assertTrue(msg.contains("[REDACTED]"));
    }

    @Test
    void redactBodyScrubsNestedSecrets() {
        Map<String, Object> body = new HashMap<>();
        body.put("user", "alice");
        body.put("password", "hunter2");
        Map<String, Object> nested = new HashMap<>();
        nested.put("api_key", "sk_xxx");
        body.put("nested", nested);
        @SuppressWarnings("unchecked")
        Map<String, Object> safe = (Map<String, Object>) Scope.redactBody(body);
        assertEquals("[REDACTED]", safe.get("password"));
        @SuppressWarnings("unchecked")
        Map<String, Object> safeNested = (Map<String, Object>) safe.get("nested");
        assertEquals("[REDACTED]", safeNested.get("api_key"));
        assertEquals("alice", safe.get("user"));
    }
}
